docs/src/main/asciidoc/writing-native-applications-tips.adoc

```diff
@@ -0,0 +1,129 @@
+
+To solve the problem shown in the example above one would simply have to declare:
+
+[source,java]
+----
+    @BuildStep
+    ReflectiveClassBuildItem reflection() {
+        // we only need to register the method for reflection, not the fields thus we pass true as the second and false as the arguments
+        return new ReflectiveClassBuildItem("com.sun.org.apache.xerces.internal.parsers.SAXParser", true, false);
+    }
+----
+
+More information about reflection in GraalVM can be found https://github.com/oracle/graal/blob/master/substratevm/REFLECTION.md[here].
+
+=== Alternative with @RegisterForReflection
+
+In application code it is often needed to pass classes to libraries that use reflection to figure out the methods/fields of those classes.
+One such example is using http://json-b.net/[JSON-B] like so:
+
+[source,java]
+----
+    public class Person {
+        private String first;
+        private String last;
+
+        public String getFirst() {
+            return first;
+        }
+
+        public void setFirst(String first) {
+            this.first = first;
+        }
+
+        public String getLast() {
+            return last;
+        }
+
+        public void setValue(String last) {
+            this.last = last;
+        }
+    }
+
+    @Path("/person")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    public class PersonResource {
+
+        private final Jsonb jsonb;
+
+        public PersonResource() {
+            jsonb = JsonbBuilder.create(new JsonbConfig());
+        }
+
+        @GET
+        public Response list() {
+            return Response.ok(jsonb.fromJson("{\"first\":  \"foo\", \"last\":  \"bar\"}", Person.class)).build();
+        }
+    }
+----
+
+If we were to use the code above, we would get an exception like the following when using the native image:
+
+[source]
+----
+ Exception handling request to /person: org.jboss.resteasy.spi.UnhandledException: javax.json.bind.JsonbException: Can't create instance of a class: class org.acme.jsonb.Person, No default constructor found
+----
+
+To get around the problem, simply annotate `Person` with `@RegisterForReflection` so Quarkus will register the class with GraalVM as a class using reflection.
+
+== Including resources
+
+By default when building a native image, GraalVM will not include any of the resources which are on the classpath into the native image it creates. Such resources that are meant to be part of the native image need to be specified explicitly.
+
+If one were to execute the native image generation process manually, the `-H:IncludeResources=` flag would be used to point to a json configuration file that specifies which resources are to be included.
+
+Quarkus eliminates the need for such a json configuration file by allowing extension authors to specify a `SubstrateResourceBuildItem`.
```

Comment:
s/json/JSON/

Is it limited to extension authors or also to application developers?

An example on how to use it as for the others would be nice.

Reply:
I thought it was limited to extension authors but now I am looking to see how application developers could achieve the same thing :)

Reply:
OK, so unless I am not seeing something in the source code, we don't provide application developer with a way to register resources, correct?

Reply:
Is adding resources to the native image functionality that we should add support for @gsmet ?

Reply:
So you can't have a `@BuildStep` in an application? I wasn't sure of that.

Yes, we would certainly need a way for applications to do all the GraalVM magic we do in extensions.

Reply:
That is something I will check, but I don't think so

Reply:
So it's technically possible, but one would have to have this kind of stuff in their own module - effectively writing a small extension.